### ISSUE TITLE
Cannot compile against 1.8.4 in Fedora 29

### ISSUE THREAD

**Issue by gone-bush** [pupnp/pupnp#107]
I cannot compile against 1.8.4 after making the required changes due to the API change. The struct UpnpFileInfo does not appear to be defined.

I am using Fedora 29 - fully up-to-date, the libupnp info is:

`libupnp-devel.x86_64	1.6.25-1.fc28`
`libupnp.x86_64			1.6.25-1.fc28`

My code, in part, is as follows.

`#include <errno.h>`
`#include <stdio.h>`
`#include <stdlib.h>`
`#include <string.h>`
`#include <unistd.h>`
`#include <sys/stat.h>`
`#include <sys/types.h>`
`#include <upnp/upnp.h>`
`...`

`int WebCallbacks::getInfo( const char* urlPath, UpnpFileInfo* info, const void *cookie )`
...
`	info->content_type  = strdup( db->getTrackMIME( db->getTrackId( 0, 0, 0, m_path, NULL )));`
`./upnp/WebCallbacks.cpp: In member function ‘int WebCallbacks::getInfo_(const char*, UpnpFileInfo*)’:`
`./upnp/WebCallbacks.cpp:98:6: error: invalid use of incomplete type ‘UpnpFileInfo’ {aka ‘struct s_UpnpFileInfo’}`
`  info->content_type  = strdup( db->getTrackMIME( db->getTrackId( 0, 0, 0, m_path, NULL )));`
`      ^~`
`In file included from /usr/include/upnp/ActionComplete.h:22,`
`                 from /usr/include/upnp/upnp.h:397,`
`                 from ./upnp/WebCallbacks.cpp:9:`
`/usr/include/upnp/TemplateInclude.h:138:1: note: forward declaration of ‘UpnpFileInfo’ {aka ‘struct s_UpnpFileInfo’}`
` TEMPLATE_PROTOTYPE_COMMON(CLASS)`
` ^~~~~~~~~~~~~~~~~~~~~~~~~`
`

I have done a "git clone" on this repository and run the following command - it doesn't find the actual definition of the struct.

`$ find . -type f -name *h -exec grep -n FileInfo {} \; -print`
`402:#include "FileInfo.h"`
`2577:    UpnpFileInfo *info,`
`./upnp/inc/upnp.h`
`7: * \brief UpnpFileInfo object declararion.`
`23:#define CLASS UpnpFileInfo`
`./upnp/inc/FileInfo.h`

**Comment by mrjimenez** [comment]
Hi,

The problem here is that you cannot dereference _info_. You must use one of the accessing functions. In this case, you should use something like this:

`
int ret;
ret = UpnpFileInfo_set_ContentType(info, db->getTrackMIME(db->getTrackId(0, 0, 0, m_path, NULL )));
`

Notice that I have removed the strdup(), that would leak memory because _set_ functions copy memory contents to their own managed buffers. By the way, avoiding those memory leaks is the reason the new API has been created.

 The key to understand this relative mess, is to look at these files:

./upnp/inc/FileInfo.h
./upnp/inc/TemplateInclude.h
./upnp/inc/TemplateSource.h

Also note that _DOMString_ is defined in ixml.h to be _char *_

Once you get used to it, it is really not that confuse.

Best regards,
Marcelo.

**Comment by gone-bush** [comment]
Thank you for your prompt and informative reply - I am now able to compile against 1.8.4.

But, and for the record, where are these setters/getters documented (I managed to guess the ones I needed)?

Cheers,
Kevin

**Comment by mrjimenez** [comment]
Hi Kevin,

Now you hit a key point: unfortunately, they are not formally documented. :( The best we can do now would be to write a text file and explain how it works and for the people to look at the appropriate include files.

When I created the templates, I tried to auto-document them using Doxygen. I believe that this work can succeed, but I currently have no time resources to dedicate to it.

All this template generated code is a C pre-processor hack. The original idea was to get the result of the pre-processor and pass it through Doxygen. If I recall correctly, the Doxygen stuff is mostly already there.

Hopefully, we can still do that one day.

Best regards,
Marcelo.
